File: spicedb-go/client/errors_test.go
package client

import (
	"context"
	"errors"
	"net"
	"testing"

	"github.com/stretchr/testify/require"
	"google.golang.org/genproto/googleapis/rpc/errdetails"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/grpc/test/bufconn"

	v1 "github.com/authzed/spicedb-clients/proto-clients/spicedb-go-proto/gen/authzed/api/v1"
	"github.com/authzed/spicedb-clients/spicedb-go/consistency"
	"github.com/authzed/spicedb-clients/spicedb-go/rel"
)

// allErrorSentinels lists every sentinel mapGRPCError can produce, used to
// assert that a mapped error matches its own sentinel and no others.
var allErrorSentinels = []error{
	ErrNotFound,
	ErrAlreadyExists,
	ErrInvalidArgument,
	ErrFailedPrecondition,
	ErrPermissionDenied,
	ErrUnauthenticated,
	ErrUnavailable,
	ErrCanceled,
	ErrDeadlineExceeded,
	ErrResourceExhausted,
	ErrOutOfRange,
}

func TestMapGRPCError_MapsKnownCodesToSentinels(t *testing.T) {
	cases := []struct {
		name       string
		code       codes.Code
		nativeCode ErrorCode
		sentinel   error
	}{
		{"NotFound", codes.NotFound, CodeNotFound, ErrNotFound},
		{"AlreadyExists", codes.AlreadyExists, CodeAlreadyExists, ErrAlreadyExists},
		{"InvalidArgument", codes.InvalidArgument, CodeInvalidArgument, ErrInvalidArgument},
		{"FailedPrecondition", codes.FailedPrecondition, CodeFailedPrecondition, ErrFailedPrecondition},
		{"PermissionDenied", codes.PermissionDenied, CodePermissionDenied, ErrPermissionDenied},
		{"Unauthenticated", codes.Unauthenticated, CodeUnauthenticated, ErrUnauthenticated},
		{"Unavailable", codes.Unavailable, CodeUnavailable, ErrUnavailable},
		{"Canceled", codes.Canceled, CodeCanceled, ErrCanceled},
		{"DeadlineExceeded", codes.DeadlineExceeded, CodeDeadlineExceeded, ErrDeadlineExceeded},
		{"ResourceExhausted", codes.ResourceExhausted, CodeResourceExhausted, ErrResourceExhausted},
		{"OutOfRange", codes.OutOfRange, CodeOutOfRange, ErrOutOfRange},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			src := status.Error(tc.code, "boom")
			got := mapGRPCError("some op", src)
			require.Error(t, got)

			require.True(t, errors.Is(got, tc.sentinel), "expected errors.Is(got, %v) to be true", tc.sentinel)

			for _, other := range allErrorSentinels {
				if other == tc.sentinel {
					continue
				}
				require.False(t, errors.Is(got, other), "expected errors.Is(got, %v) to be false", other)
			}

			nativeErr, ok := got.(*Error)
			require.True(t, ok, "expected mapGRPCError to return *Error")
			require.Equal(t, tc.nativeCode, nativeErr.Code)
			require.Contains(t, nativeErr.Message, "some op")
			require.Contains(t, nativeErr.Message, "boom")

			require.NotNil(t, errors.Unwrap(got))
		})
	}
}

func TestMapGRPCError_NilErrorReturnsNil(t *testing.T) {
	got := mapGRPCError("x", nil)
	require.Nil(t, got)
	require.True(t, got == nil)
}

func TestMapGRPCError_UnmappedCodeMatchesNoSentinel(t *testing.T) {
	src := status.Error(codes.Internal, "boom")
	got := mapGRPCError("some op", src)
	require.Error(t, got)

	nativeErr, ok := got.(*Error)
	require.True(t, ok, "expected mapGRPCError to return *Error")
	require.Equal(t, CodeInternal, nativeErr.Code)

	for _, sentinel := range allErrorSentinels {
		require.False(t, errors.Is(got, sentinel), "codes.Internal should not match sentinel %v", sentinel)
	}
}

func TestMapGRPCError_UnknownCodeMapsToCodeUnknown(t *testing.T) {
	// codes.Code(9999) has no entry in the internal codes.Code -> ErrorCode
	// mapping table, so it must fall back to CodeUnknown rather than the zero
	// value of some unrelated code.
	src := status.Error(codes.Code(9999), "mystery")
	got := mapGRPCError("some op", src)
	require.Error(t, got)

	nativeErr, ok := got.(*Error)
	require.True(t, ok, "expected mapGRPCError to return *Error")
	require.Equal(t, CodeUnknown, nativeErr.Code)
}

// erroringPermissionsServer's ReadRelationships fails immediately with
// codes.NotFound, simulating a server-side error surfaced mid-stream.
type erroringPermissionsServer struct {
	v1.UnimplementedPermissionsServiceServer
}

func (s *erroringPermissionsServer) ReadRelationships(_ *v1.ReadRelationshipsRequest, _ grpc.ServerStreamingServer[v1.ReadRelationshipsResponse]) error {
	return status.Error(codes.NotFound, "nope")
}

// startErroringServer starts an in-process gRPC server backed by bufconn,
// exposing a PermissionsService whose ReadRelationships always fails with
// codes.NotFound, and returns a dialer for it.
func startErroringServer(t *testing.T) func(context.Context, string) (net.Conn, error) {
	t.Helper()

	const bufSize = 1024 * 1024
	lis := bufconn.Listen(bufSize)

	grpcServer := grpc.NewServer()
	v1.RegisterPermissionsServiceServer(grpcServer, &erroringPermissionsServer{})

	go func() {
		_ = grpcServer.Serve(lis)
	}()
	t.Cleanup(grpcServer.Stop)

	return func(ctx context.Context, _ string) (net.Conn, error) {
		return lis.DialContext(ctx)
	}
}

// TestReadRelationships_StreamErrorIsMappedToNativeError proves that errors
// surfaced mid-stream (i.e. from stream.Recv(), not just the initial RPC
// call) are routed through mapGRPCError, so iterator consumers can use
// errors.Is against the native sentinels instead of inspecting raw gRPC
// status codes.
func TestReadRelationships_StreamErrorIsMappedToNativeError(t *testing.T) {
	dialer := startErroringServer(t)

	c, err := NewWithOpts("passthrough:///bufnet", "test-token",
		WithInsecure(),
		// bufnet is an in-memory bufconn dial target, not a real network
		// destination -- unrelated to what this test exercises.
		WithInsecureAllowRemoteHost(),
		WithDialOptions(grpc.WithContextDialer(dialer)),
	)
	require.NoError(t, err)

	var (
		gotErr error
		yields int
	)
	for _, iterErr := range c.ReadRelationships(context.Background(), consistency.MinLatency(), rel.NewFilter("document")) {
		yields++
		if iterErr != nil {
			gotErr = iterErr
		}
	}

	require.Equal(t, 1, yields, "expected exactly one yield carrying the mapped error")
	require.Error(t, gotErr)
	require.True(t, errors.Is(gotErr, ErrNotFound), "expected the stream error to satisfy errors.Is(err, client.ErrNotFound), got: %v", gotErr)

	nativeErr, ok := gotErr.(*Error)
	require.True(t, ok, "expected the stream error to be a native *Error")
	require.Equal(t, CodeNotFound, nativeErr.Code)
}

// TestErrorCode_IsNativeType is a compile-time-flavored assertion that
// Error.Code is the native ErrorCode type, not grpc/codes.Code: assigning it
// to an ErrorCode-typed variable (with no conversion) would fail to compile
// if the field's type ever regressed to codes.Code.
func TestErrorCode_IsNativeType(t *testing.T) {
	var e Error
	var code ErrorCode = e.Code //nolint:staticcheck // intentional: explicit type is the compile-time guard described above
	require.Equal(t, CodeUnknown, code)
}

// TestMapGRPCError_SurfacesErrorReason proves that SpiceDB's structured
// explanation for a failure -- the google.rpc.ErrorInfo detail carried on the
// server's status -- survives the mapping into a native *Error, so a caller
// can branch on the reason and read its metadata instead of string-matching
// the message. See root DESIGN.md, "Error mapping must not lose the server's
// detail".
func TestMapGRPCError_SurfacesErrorReason(t *testing.T) {
	st, err := status.New(codes.ResourceExhausted, "max depth exceeded").WithDetails(&errdetails.ErrorInfo{
		Reason:   v1.ErrorReason_ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED.String(),
		Domain:   "authzed.com",
		Metadata: map[string]string{"maximum_depth_allowed": "50"},
	})
	require.NoError(t, err)

	got := mapGRPCError("check", st.Err())
	nativeErr, ok := got.(*Error)
	require.True(t, ok, "expected mapGRPCError to return *Error")

	// The exposed reason is exactly the authzed.api.v1.ErrorReason enum name,
	// so a caller can compare against the generated enum without this client
	// carrying a hand-maintained copy of it.
	require.Equal(t, "ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED", nativeErr.Reason)
	require.Equal(t, v1.ErrorReason_ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED.String(), nativeErr.Reason)
	require.Equal(t, "authzed.com", nativeErr.ReasonDomain)
	require.Equal(t, map[string]string{"maximum_depth_allowed": "50"}, nativeErr.ReasonMetadata)
}

// TestMapGRPCError_PreconditionFailureReasonMetadata proves the metadata map
// is not dropped: a caller must be able to read *which* precondition failed,
// not merely that one did.
func TestMapGRPCError_PreconditionFailureReasonMetadata(t *testing.T) {
	st, err := status.New(codes.FailedPrecondition, "precondition failed").WithDetails(&errdetails.ErrorInfo{
		Reason: v1.ErrorReason_ERROR_REASON_WRITE_OR_DELETE_PRECONDITION_FAILURE.String(),
		Domain: "authzed.com",
		Metadata: map[string]string{
			"precondition_operation":     "OPERATION_MUST_MATCH",
			"precondition_resource_type": "document",
			"precondition_resource_id":   "firstdoc",
			"precondition_relation":      "viewer",
		},
	})
	require.NoError(t, err)

	nativeErr, ok := mapGRPCError("write", st.Err()).(*Error)
	require.True(t, ok)
	require.Equal(t, "ERROR_REASON_WRITE_OR_DELETE_PRECONDITION_FAILURE", nativeErr.Reason)
	require.Equal(t, "firstdoc", nativeErr.ReasonMetadata["precondition_resource_id"])
	require.Equal(t, "viewer", nativeErr.ReasonMetadata["precondition_relation"])
}

// TestMapGRPCError_UnrecognizedReasonPassesThrough proves an ErrorReason a
// newer server knows and this client does not is surfaced verbatim rather
// than raising or being coerced. The reason is server-supplied, so root
// DESIGN.md's "A conversion that cannot preserve meaning must fail" requires
// it to degrade safely, not to fail.
func TestMapGRPCError_UnrecognizedReasonPassesThrough(t *testing.T) {
	st, err := status.New(codes.InvalidArgument, "from the future").WithDetails(&errdetails.ErrorInfo{
		Reason:   "ERROR_REASON_INVENTED_BY_A_NEWER_SERVER",
		Domain:   "authzed.com",
		Metadata: map[string]string{"k": "v"},
	})
	require.NoError(t, err)

	nativeErr, ok := mapGRPCError("check", st.Err()).(*Error)
	require.True(t, ok)
	require.Equal(t, CodeInvalidArgument, nativeErr.Code)
	require.Equal(t, "ERROR_REASON_INVENTED_BY_A_NEWER_SERVER", nativeErr.Reason)
	require.Equal(t, map[string]string{"k": "v"}, nativeErr.ReasonMetadata)
}

// TestMapGRPCError_NoErrorInfoLeavesReasonEmpty proves the absence of an
// ErrorInfo detail is represented as an empty reason, not a fabricated one.
func TestMapGRPCError_NoErrorInfoLeavesReasonEmpty(t *testing.T) {
	nativeErr, ok := mapGRPCError("check", status.Error(codes.NotFound, "nope")).(*Error)
	require.True(t, ok)
	require.Empty(t, nativeErr.Reason)
	require.Empty(t, nativeErr.ReasonDomain)
	require.Empty(t, nativeErr.ReasonMetadata)
}

// TestMapGRPCError_OutOfRangeIsExpiredZedToken documents what OUT_OF_RANGE
// means in SpiceDB -- an expired or garbage-collected ZedToken -- and proves
// it is distinguishable from every other mapped code, so the mechanical
// recovery (drop the token, re-read at full consistency) can be written
// against a type rather than a message.
func TestMapGRPCError_OutOfRangeIsExpiredZedToken(t *testing.T) {
	got := mapGRPCError("check", status.Error(codes.OutOfRange, "revision is no longer available"))
	require.True(t, errors.Is(got, ErrOutOfRange))
	require.False(t, errors.Is(got, ErrInvalidArgument))
	require.False(t, errors.Is(got, ErrFailedPrecondition))

	nativeErr, ok := got.(*Error)
	require.True(t, ok)
	require.Equal(t, CodeOutOfRange, nativeErr.Code)
	require.Equal(t, "OutOfRange", nativeErr.Code.String())
}

// TestMapGRPCError_UnauthenticatedIsNotInternal proves a rotated or wrong API
// token is distinguishable from an internal server fault, so a caller can
// refresh credentials on one and page someone on the other.
func TestMapGRPCError_UnauthenticatedIsNotInternal(t *testing.T) {
	got := mapGRPCError("check", status.Error(codes.Unauthenticated, "bad token"))
	require.True(t, errors.Is(got, ErrUnauthenticated))

	nativeErr, ok := got.(*Error)
	require.True(t, ok)
	require.Equal(t, CodeUnauthenticated, nativeErr.Code)
	require.NotEqual(t, CodeInternal, nativeErr.Code)
}

File: spicedb-typescript/src/__tests__/errors.test.ts
import { describe, it, expect } from "vitest";
import { ConnectError, Code } from "@connectrpc/connect";
import { create, toBinary } from "@bufbuild/protobuf";
import { anyPack } from "@bufbuild/protobuf/wkt";
import { ErrorInfoSchema, ErrorReason, ErrorReasonSchema } from "@spicedb/proto";
import {
  SpiceDBError,
  PermissionDeniedError,
  NotFoundError,
  AlreadyExistsError,
  InvalidArgumentError,
  UnavailableError,
  UnauthenticatedError,
  OutOfRangeError,
  FailedPreconditionError,
  CancelledError,
  DeadlineExceededError,
  ResourceExhaustedError,
  toSpiceDBError,
  toSpiceDBErrorFromStatus,
  isTransientError,
} from "../errors.js";

/**
 * Builds a ConnectError carrying a `google.rpc.ErrorInfo` detail in the shape
 * the gRPC transport produces for one: an incoming detail is the type name
 * plus the serialized message, not a schema/value pair.
 */
function connectErrorWithReason(
  code: Code,
  message: string,
  reason: string,
  metadata: Record<string, string>,
  domain = "authzed.com",
): ConnectError {
  const err = new ConnectError(message, code);
  err.details = [
    {
      type: ErrorInfoSchema.typeName,
      value: toBinary(
        ErrorInfoSchema,
        create(ErrorInfoSchema, { reason, domain, metadata }),
      ),
    },
  ];
  return err;
}

describe("toSpiceDBError", () => {
  it("maps PermissionDenied", () => {
    const err = new ConnectError("denied", Code.PermissionDenied);
    const result = toSpiceDBError(err);
    expect(result).toBeInstanceOf(PermissionDeniedError);
    expect(result.cause).toBe(err);
  });

  it("maps NotFound", () => {
    const err = new ConnectError("not found", Code.NotFound);
    expect(toSpiceDBError(err)).toBeInstanceOf(NotFoundError);
  });

  it("maps AlreadyExists", () => {
    const err = new ConnectError("exists", Code.AlreadyExists);
    expect(toSpiceDBError(err)).toBeInstanceOf(AlreadyExistsError);
  });

  it("maps InvalidArgument", () => {
    const err = new ConnectError("bad", Code.InvalidArgument);
    expect(toSpiceDBError(err)).toBeInstanceOf(InvalidArgumentError);
  });

  it("maps FailedPrecondition", () => {
    const err = new ConnectError("precondition", Code.FailedPrecondition);
    expect(toSpiceDBError(err)).toBeInstanceOf(FailedPreconditionError);
  });

  it("maps Unavailable", () => {
    const err = new ConnectError("unavailable", Code.Unavailable);
    expect(toSpiceDBError(err)).toBeInstanceOf(UnavailableError);
  });

  it("maps Cancelled", () => {
    const err = new ConnectError("cancelled", Code.Canceled);
    expect(toSpiceDBError(err)).toBeInstanceOf(CancelledError);
  });

  it("maps DeadlineExceeded", () => {
    const err = new ConnectError("timeout", Code.DeadlineExceeded);
    expect(toSpiceDBError(err)).toBeInstanceOf(DeadlineExceededError);
  });

  it("maps ResourceExhausted", () => {
    const err = new ConnectError("quota", Code.ResourceExhausted);
    expect(toSpiceDBError(err)).toBeInstanceOf(ResourceExhaustedError);
  });

  it("maps unknown errors", () => {
    const err = new ConnectError("unknown", Code.Internal);
    expect(toSpiceDBError(err)).toBeInstanceOf(SpiceDBError);
  });

  it("wraps non-ConnectError", () => {
    const err = new Error("boom");
    const result = toSpiceDBError(err);
    expect(result).toBeInstanceOf(SpiceDBError);
    expect(result.message).toBe("boom");
  });

  it("wraps non-Error", () => {
    const result = toSpiceDBError("oops");
    expect(result).toBeInstanceOf(SpiceDBError);
    expect(result.message).toBe("oops");
  });

  it("passes through SpiceDBError", () => {
    const err = new NotFoundError("nope");
    expect(toSpiceDBError(err)).toBe(err);
  });

  // OUT_OF_RANGE is SpiceDB's code for an expired or garbage-collected
  // ZedToken. Recovery is mechanical -- drop the token, re-read at full
  // consistency -- so it has to be distinguishable by type rather than by
  // message.
  it("maps OutOfRange to its own type", () => {
    const err = new ConnectError("revision no longer available", Code.OutOfRange);
    const result = toSpiceDBError(err);
    expect(result).toBeInstanceOf(OutOfRangeError);
    expect(result).toBeInstanceOf(SpiceDBError);
    expect(result).not.toBeInstanceOf(InvalidArgumentError);
    expect(result.name).toBe("OutOfRangeError");
  });

  // A wrong, expired, or rotated token must be distinguishable from an
  // internal server fault, so a caller can refresh credentials on one and page
  // someone on the other.
  it("maps Unauthenticated to its own type", () => {
    const err = new ConnectError("bad token", Code.Unauthenticated);
    const result = toSpiceDBError(err);
    expect(result).toBeInstanceOf(UnauthenticatedError);
    expect(result).not.toBeInstanceOf(PermissionDeniedError);
    expect(result.name).toBe("UnauthenticatedError");
  });

  it("neither newly mapped code is retried", () => {
    expect(isTransientError(new ConnectError("x", Code.OutOfRange))).toBe(false);
    expect(isTransientError(new ConnectError("x", Code.Unauthenticated))).toBe(
      false,
    );
  });
});

// SpiceDB's structured explanation of a failure -- the google.rpc.ErrorInfo
// detail on the status -- must survive the mapping into a typed error, so a
// caller can branch on the reason and read its metadata instead of
// string-matching a message. See root DESIGN.md, "Error mapping must not lose
// the server's detail".
describe("ErrorReason is reachable", () => {
  it("surfaces the reason, domain, and metadata", () => {
    const err = connectErrorWithReason(
      Code.ResourceExhausted,
      "max depth exceeded",
      "ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED",
      { maximum_depth_allowed: "50" },
    );
    const result = toSpiceDBError(err);

    expect(result).toBeInstanceOf(ResourceExhaustedError);
    // The exposed reason is exactly the authzed.api.v1.ErrorReason enum name,
    // so a caller can compare against the generated enum without this client
    // carrying a hand-maintained copy of it.
    expect(result.reason).toBe("ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED");
    expect(result.reason).toBe(
      ErrorReasonSchema.value[ErrorReason.MAXIMUM_DEPTH_EXCEEDED]?.name,
    );
    expect(result.reasonDomain).toBe("authzed.com");
    expect(result.reasonMetadata).toEqual({ maximum_depth_allowed: "50" });
  });

  it("keeps the metadata naming which precondition failed", () => {
    const err = connectErrorWithReason(
      Code.FailedPrecondition,
      "precondition failed",
      "ERROR_REASON_WRITE_OR_DELETE_PRECONDITION_FAILURE",
      {
        precondition_resource_id: "firstdoc",
        precondition_relation: "viewer",
      },
    );
    const result = toSpiceDBError(err);

    expect(result).toBeInstanceOf(FailedPreconditionError);
    expect(result.reason).toBe(
      "ERROR_REASON_WRITE_OR_DELETE_PRECONDITION_FAILURE",
    );
    expect(result.reasonMetadata.precondition_resource_id).toBe("firstdoc");
    expect(result.reasonMetadata.precondition_relation).toBe("viewer");
  });

  // A reason a newer server knows and this client does not is server-supplied:
  // root DESIGN.md's "A conversion that cannot preserve meaning must fail"
  // requires it to degrade safely, not to throw.
  it("passes an unrecognized reason through without throwing", () => {
    const err = connectErrorWithReason(
      Code.InvalidArgument,
      "from the future",
      "ERROR_REASON_INVENTED_BY_A_NEWER_SERVER",
      { k: "v" },
    );
    const result = toSpiceDBError(err);

    expect(result).toBeInstanceOf(InvalidArgumentError);
    expect(result.reason).toBe("ERROR_REASON_INVENTED_BY_A_NEWER_SERVER");
    expect(result.reasonMetadata).toEqual({ k: "v" });
  });

  it("leaves the reason empty when the server attached no ErrorInfo", () => {
    const result = toSpiceDBError(new ConnectError("nope", Code.NotFound));
    expect(result.reason).toBe("");
    expect(result.reasonDomain).toBe("");
    expect(result.reasonMetadata).toEqual({});
  });

  // A per-item bulk error arrives as a google.rpc.Status with its own details,
  // and must not lose them on the way to a typed error.
  it("surfaces the reason from a per-item google.rpc.Status", () => {
    const result = toSpiceDBErrorFromStatus({
      code: 8,
      message: "max depth exceeded",
      details: [
        anyPack(
          ErrorInfoSchema,
          create(ErrorInfoSchema, {
            reason: "ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED",
            domain: "authzed.com",
            metadata: { maximum_depth_allowed: "50" },
          }),
        ),
      ],
    });

    expect(result).toBeInstanceOf(ResourceExhaustedError);
    expect(result.reason).toBe("ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED");
    expect(result.reasonMetadata).toEqual({ maximum_depth_allowed: "50" });
  });
});

describe("toSpiceDBErrorFromStatus", () => {
  // Regression coverage for the CheckBulkPermissions per-item error path:
  // a google.rpc.Status-shaped per-item error (numeric code, not a
  // ConnectError instance) must map to the same typed error classes as a
  // top-level RPC failure — a caller must not be able to tell a per-item
  // error apart from an RPC-level one except by catching a typed error.
  it("maps PermissionDenied (code 7)", () => {
    const result = toSpiceDBErrorFromStatus({ code: 7, message: "nope" });
    expect(result).toBeInstanceOf(PermissionDeniedError);
  });

  it("maps InvalidArgument (code 3)", () => {
    const result = toSpiceDBErrorFromStatus({ code: 3, message: "bad" });
    expect(result).toBeInstanceOf(InvalidArgumentError);
  });

  it("maps an unrecognized/internal code to the base SpiceDBError", () => {
    const result = toSpiceDBErrorFromStatus({ code: 13, message: "boom" });
    expect(result).toBeInstanceOf(SpiceDBError);
  });

  it("carries the status message through", () => {
    const result = toSpiceDBErrorFromStatus({ code: 5, message: "no such thing" });
    expect(result).toBeInstanceOf(NotFoundError);
    expect(result.message).toContain("no such thing");
  });
});

describe("isTransientError", () => {
  it("returns true for Unavailable", () => {
    expect(
      isTransientError(new ConnectError("", Code.Unavailable)),
    ).toBe(true);
  });

  it("returns false for ResourceExhausted", () => {
    // Inverted from "returns true" -- RESOURCE_EXHAUSTED must NOT be
    // retried. In SpiceDB it signals memory load-shed (retrying adds load
    // to an already-overloaded server) or a deterministic MaxDepthExceeded
    // (retrying can never succeed). See DESIGN.md, "Automatic retry is for
    // idempotent operations only".
    expect(
      isTransientError(new ConnectError("", Code.ResourceExhausted)),
    ).toBe(false);
  });

  it("returns true for Aborted", () => {
    expect(
      isTransientError(new ConnectError("", Code.Aborted)),
    ).toBe(true);
  });

  it("returns false for PermissionDenied", () => {
    expect(
      isTransientError(new ConnectError("", Code.PermissionDenied)),
    ).toBe(false);
  });

  it("returns false for non-ConnectError", () => {
    expect(isTransientError(new Error("boom"))).toBe(false);
  });
});
